### PR TITLE
Initial implementation of amazon dynamodb siddhi store

### PR DESCRIPTION
**Pull request by ShevindiRodrigo**
## Purpose
Amazon DynamoDB connector will make it possible for users to insert, update, update or insert data into a dynamodb table and reading and deleting data from a dynamodb table.

## Goals
This connector allows user to store data in a Amazon DynamoDB database.

## Approach
Amazon DynamoDB is used for this connector.

## User stories
https://docs.google.com/document/d/149uFt9USWVV1vV2xbAv4-3R46tm_667LSqRYYfiMtDQ/edit#heading=h.ymxvja9dvl0w

## Automation tests
 - Unit tests 
   78%
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Test environment
JDK versions : JDK 8
operating systems : ubuntu 16.04 LTS
databases : amazon dynamodb

### CODE REVIEW

File: component/pom.xml [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.module</groupId>
+            <artifactId>jackson-module-osgi</artifactId>
+        </dependency>
+    </dependencies>
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.felix</groupId>
```
Comment (niveathika) [inline component/pom.xml]: Remove duplicated plugin dev

File: component/pom.xml [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+                <extensions>true</extensions>
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Bundle-Name>${project.artifactId}</Bundle-Name>
+                        <Export-Package>
+                            org.wso2.extension.siddhi.store.amazondynamodb.*,
+                        </Export-Package>
+                        <Import-Package>
+                            *;resolution:=optional
```
Comment (niveathika) [inline component/pom.xml]: Explicitly imprort packages with version range

File: component/src/main/java/org/wso2/extension/siddhi/store/amazondynamodb/AmazonDynamoDBIterator.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        } else {
+            this.getAttributeValues = scanResult.getItems();
+        }
+    }
+
+    private static Object getValueFromAttributeValue(AttributeValue value) {
+        if (value == null) {
+            return null;
+        }
+        if (value.getN() != null) {
```
Comment (niveathika) [inline component/src/main/java/org/wso2/extension/siddhi/store/amazondynamodb/AmazonDynamoDBIterator.java]: Use switch case

File: component/src/main/java/org/wso2/extension/siddhi/store/amazondynamodb/AmazonDynamoDBStore.java [hunk 4]
```diff
@@ -0,0 +1,1163 @@
...
+                    .withTableName(this.tableName)
+                    .withItem(attributesValueMap);
+            client.putItem(putItemRequest);
+        } catch (AmazonDynamoDBException e) {
+            throw new AmazonDynamoDBTableException("Error while inserting records to " + this.tableName + " : "
+                    + e.getMessage());
+        }
+    }
+
+    /**
```
Comment (niveathika) [inline component/src/main/java/org/wso2/extension/siddhi/store/amazondynamodb/AmazonDynamoDBStore.java]: Remove inherited javadocs

File: component/src/test/java/org/wso2/extension/siddhi/store/amazondynamodb/AmazonDynamoDBStoreQuery.java [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+import org.wso2.siddhi.query.compiler.exception.SiddhiParserException;
+
+import static org.wso2.extension.siddhi.store.amazondynamodb.AmazonDynamoDBTestUtils.ACCESS_KEY;
+import static org.wso2.extension.siddhi.store.amazondynamodb.AmazonDynamoDBTestUtils.READ_CAPACITY_UNITS;
+import static org.wso2.extension.siddhi.store.amazondynamodb.AmazonDynamoDBTestUtils.SECRET_KEY;
+import static org.wso2.extension.siddhi.store.amazondynamodb.AmazonDynamoDBTestUtils.SIGNING_REGION;
+import static org.wso2.extension.siddhi.store.amazondynamodb.AmazonDynamoDBTestUtils.TABLE_NAME;
+import static org.wso2.extension.siddhi.store.amazondynamodb.AmazonDynamoDBTestUtils.WRITE_CAPACITY_UNITS;
+
+public class AmazonDynamoDBStoreQuery {
```
Comment (niveathika) [inline component/src/test/java/org/wso2/extension/siddhi/store/amazondynamodb/AmazonDynamoDBStoreQuery.java]: Rename testcases to end with "TestCase"

File: docs/images/siddhi-logo-w.svg [hunk 6]
```diff
@@ -0,0 +1,163 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<!--
+  ~ Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (niveathika) [inline docs/images/siddhi-logo-w.svg]: License year, check in other places as well

File: mkdocs.yml [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  palette:
+    primary: teal
+    accent: teal
+google_analytics:
+  - UA-103065-28
+  - auto
+extra:
+  social:
+    - type: github
+      link: https://github.com/wso2/siddhi
```
Comment (niveathika) [inline mkdocs.yml]: Point to new siddhi repo

File: pom.xml [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+    <groupId>org.wso2.extension.siddhi.store.amazondynamodb</groupId>
+    <artifactId>siddhi-store-amazondynamodb-parent</artifactId>
+    <version>1.0.0</version>
+    <name>WSO2 Siddhi Extension - amazonDynamoDB Store</name>
+    <modelVersion>4.0.0</modelVersion>
+    <packaging>pom</packaging>
+    <modules>
+        <module>component</module>
+    </modules>
+    <profiles>
```
Comment (niveathika) [inline pom.xml]: Move profiles, to component pom. else doc won't be generated

File: pom.xml [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-release-plugin</artifactId>
+                <configuration>
+                    <preparationGoals>clean install -Pdocumentation-deploy</preparationGoals>
+                    <autoVersionSubmodules>true</autoVersionSubmodules>
+                </configuration>
+            </plugin>
+        </plugins>
+        <pluginManagement>
```
Comment (niveathika) [inline pom.xml]: Remove as these are not used

File: component/src/main/java/org/wso2/extension/siddhi/store/amazondynamodb/AmazonDynamoDBIterator.java [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    public Object[] next() {
+        if (this.hasNext()) {
+            return extractRecord(queryResult, scanResult);
+        }
+        return new Object[0];
+    }
+
+    private Object[] extractRecord(QueryResult queryResult, ScanResult scanResult) {
+        List<Object> objects = new ArrayList<>();
+
```
Comment (ramindu90) [inline component/src/main/java/org/wso2/extension/siddhi/store/amazondynamodb/AmazonDynamoDBIterator.java]: Please remove unnecessary spaces in all files

File: component/src/test/java/org/wso2/extension/siddhi/store/amazondynamodb/ContainsAmazonDynamoDBTestCase.java [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+            inEventCount = 0;
+            eventArrived = false;
+            AmazonDynamoDBTestUtils.init(TABLE_NAME);
+        } catch (AmazonDynamoDBException e) {
+            log.info("Test case ignored due to :" + e.getMessage());
+        }
+    }
+
+
+    @Test(description = "Test contains with one condition.")
```
[the remark this answers is not in the document]
Comment (ShevindiRodrigo) [reply component/src/test/java/org/wso2/extension/siddhi/store/amazondynamodb/ContainsAmazonDynamoDBTestCase.java]: Yes we can run without making any changes.

File: component/src/main/java/org/wso2/extension/siddhi/store/amazondynamodb/AmazonDynamoDBStore.java [hunk 12]
```diff
@@ -0,0 +1,1163 @@
...
+                        .withKeySchema(elements)
+                        .withProvisionedThroughput(new ProvisionedThroughput()
+                                .withReadCapacityUnits(readCapacityUnits)
+                                .withWriteCapacityUnits(writeCapacityUnits))
+                        .withAttributeDefinitions(attributeDefinitions);
+                Table table = dynamoDB.createTable(createTableRequest);
+                table.waitForActive();
+            }
+        } catch (AmazonDynamoDBTableException | InterruptedException e) {
+            throw new ConnectionUnavailableException("Failed to create table " + this.tableName + " due to : "
```
[the remark this answers is not in the document]
Comment (ShevindiRodrigo) [reply component/src/main/java/org/wso2/extension/siddhi/store/amazondynamodb/AmazonDynamoDBStore.java]: Yes it is recoverable by retrying.

**Pull request merged**